Python/Product/PythonTools/PythonTools/Environments/AddCondaEnvironmentView.cs

```diff
@@ -0,0 +1,445 @@
+            }
+        }
+
+        private async Task UpdatePreviewAsync(string envName, string envFilePath, CancellationToken ct) {
+            try {
+                ct.ThrowIfCancellationRequested();
+                string result;
+                try {
+                    result = File.ReadAllText(envFilePath);
+                } catch (UnauthorizedAccessException ex) {
+                    result = ex.Message;
+                } catch (ArgumentException ex) {
+                    result = ex.Message;
+                } catch (FileNotFoundException ex) {
+                    result = ex.Message;
+                } catch (IOException ex) {
+                    result = ex.Message;
+                }
+                ct.ThrowIfCancellationRequested();
+                Dispatcher.Invoke(() => {
+                    var p = new CondaEnvironmentPreview();
+                    p.IsInProgress = false;
+                    p.IsEnvFile = result != null;
+                    p.IsNoEnvFile = result == null;
+                    p.EnvFileContents = result ?? "";
+                    CondaPreview = p;
+                });
+            } catch (OperationCanceledException) {
+            }
+        }
+
+        private async Task UpdatePreviewAsync(string envName, PackageSpec[] specs, CancellationToken ct) {
+            if (specs.Length == 0) {
+                Dispatcher.Invoke(() => {
+                    CondaPreview = new CondaEnvironmentPreview() {
+                        IsNoPackages = true
+                    };
+                });
+                return;
+            }
+
+            try {
+                ct.ThrowIfCancellationRequested();
+                var result = await _condaMgr.PreviewCreateAsync(envName, specs, ct);
+                ct.ThrowIfCancellationRequested();
+                var preview = result?.Actions?.LinkPackages;
+                var msg = result?.Message;
+                Dispatcher.Invoke(() => {
+                    var p = new CondaEnvironmentPreview();
+                    if (preview != null) {
+                        foreach (var package in preview.MaybeEnumerate().OrderBy(pkg => pkg.Name)) {
+                            p.Packages.Add(new CondaPackageView() {
+                                Name = package.Name.Trim(),
+                                Version = package.VersionText,
+                            });
+                        }
+                    }
+
+                    if (result == null) {
+                        msg = Strings.AddCondaEnvironmentPreviewFailed;
+                    }
+                    p.HasPreviewError = !string.IsNullOrEmpty(msg);
+                    p.ErrorMessage = msg;
+                    p.IsInProgress = false;
+                    p.IsPackages = p.Packages.Count > 0;
+                    p.IsNoPackages = !p.IsPackages && !p.HasPreviewError;
+
+                    CondaPreview = p;
+                });
+            } catch (OperationCanceledException) {
```

Comment:
If you catch OCE here, calling async method will continue execution. Is that expected?

Reply:
The calling method is a timer callback, and runs this method on thread pool and never waits on it. So that is fine.